pkg/services/ngalert/state/persister_async.go

```diff
@@ -0,0 +1,62 @@
+package state
+
+import (
+	"context"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+type AsyncStatePersister struct {
+	log log.Logger
+	// doNotSaveNormalState controls whether eval.Normal state is persisted to the database and returned by get methods.
+	doNotSaveNormalState bool
+	store                InstanceStore
+}
+
+func NewAsyncStatePersisiter(log log.Logger, cfg ManagerCfg) StatePersister {
+	return &AsyncStatePersister{
+		log:                  log,
+		store:                cfg.InstanceStore,
+		doNotSaveNormalState: cfg.DoNotSaveNormalState,
+	}
+}
+
+func (a *AsyncStatePersister) Async(ctx context.Context, ticker *clock.Ticker, cache *cache) {
+infLoop:
+	for {
+		select {
+		case <-ticker.C:
+			if err := a.fullSync(ctx, cache); err != nil {
+				a.log.Error("Failed to do a full state sync to database", "err", err)
+			}
+		case <-ctx.Done():
+			a.log.Info("Stopping state sync...")
+			if err := a.fullSync(context.Background(), cache); err != nil {
```

Comment:
If I remember in the original pull request there was some a about not doing this on shutdown because it might exceed the termination grace period in K8s and end up in the pod being sigtermed. What was the outcome of that discussion as I can't remember 😞

Reply:
Any transaction affected by this would just get timed out by the database, so I think it's fine to attempt? Is the pod getting sigterm'd a super bad thing? The worst that would happen is we fail to persist state, which is the same effect as not saving at all here.

Reply:
Agree with Alex.